src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            while (true)
+            {
+                ReadResult result = await _requestPipe.Reader.ReadAsync(cancellationToken);
+
+                if (result.IsCanceled)
+                {
+                    throw new OperationCanceledException("The read was canceled");
+                }
+
+                ReadOnlySequence<byte> buffer = result.Buffer;
+                long length = buffer.Length;
+
+                SequencePosition consumed = buffer.End;
+                try
+                {
+                    if (length != 0)
+                    {
+                        int actual = (int)Math.Min(length, destination.Length);
+
+                        ReadOnlySequence<byte> slice = actual == length ? buffer : buffer.Slice(0, actual);
+                        consumed = slice.End;
+                        slice.CopyTo(destination.Span);
+
+                        if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+                        return actual;
+                    }
+
+                    if (result.IsCompleted)
+                    {
+                        if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    _requestPipe.Reader.AdvanceTo(consumed);
+                }
+            }
+        }
+
+        internal override void ShutdownRead()
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            _requestPipe.Reader.Complete();
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+        }
+
+        internal override void ShutdownWrite()
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            // TODO do anything to stop writes?
+            _api.StreamShutdownDelegate(_ptr, (uint)QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+        }
+
+        internal override int Read(Span<byte> buffer)
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override void Write(ReadOnlySpan<byte> buffer)
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override void Flush()
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override async Task FlushAsync(CancellationToken cancellationToken)
+        {
+            await SendAsync(new ReadOnlySequence<byte>(Memory<byte>.Empty), QUIC_SEND_FLAG.NONE);
+        }
+
+        public override async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_ptr != IntPtr.Zero)
+            {
+                await ShutdownAsync(QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0).ConfigureAwait(false);
```

Comment:
I don't like implicitly picking a shutdown methodology for the app here. Personally, I'd think abortive shutdown better, but then the app *MUST* supply an error code or the entire application protocol can/will break down and have unexpected side effects.

Reply:
Yeah I need to fix this, right now I changed Dispose to just call Close without calling Abortive Shutdown.

Reply:
Is there documentation on what error codes I should supply here?